src/states/elder/mod.rs

```diff
@@ -1601,6 +1619,20 @@ impl Approved for Elder {
         // go to the new instance.
         let relocate_details = self.chain.poll_relocation();
 
+        for pub_id in &elders_change.own_added {
+            self.send_event(Event::NodeAdded(*pub_id.name()), outbox);
+        }
+
+        for pub_id in &elders_change.own_removed {
+            self.send_event(Event::NodeLost(*pub_id.name()), outbox);
+        }
+
+        if elders_info.prefix().matches(self.name())
+            && !elders_info.is_member(&self.full_id.public_id())
+        {
+            return Ok(Transition::Demote);
```

Comment:
here is something weird in handle_section_info_event:
Again, this happen only on one side of the prefix, so we may not get demotion one side of the prefix change,

        if elders_info.prefix().matches(self.name())
            && !elders_info.is_member(&self.full_id.public_id())
        {
            return Ok(Transition::Demote);
        }

Probably what we want is check we are still an elder from withing that else, checking against our_info()

 else {
            self.reset_parsec()?;
        }

Reply:
Or maybe better as:

```

        } else if !self.our_prefix().matches(self.name()) {
            return Ok(Transition::Demote);
        } else {
            self.reset_parsec()?;
        }
```

Reply:
No, that probably made more sense as a stand alone if as you did, but still should use `self.our_prefix()` instead of elder_info

Reply:
Actually, the point here is to use `elder_info`. `self.our_prefix()` will always match our name, so that wouldn't do anything.

When a split is happening, both `elder_info` are handled by `handle_section_info_event`. But even if we are supposed to remain an Elder of one of the child sections, only one of the `elder_info`s will contain our ID - the one with prefix matching our name. So when we are handling the other one, we shouldn't demote ourselves even though it won't contain our ID. We should only demote ourselves if the matching one doesn't contain it.

I hope that makes sense.

Reply:
> When a split is happening, both elder_infos are handled by handle_section_info_event.

This is not true, only one is, and it can be either side. (the other one is cached, and never passed on.

Reply:
Are you sure? Because I'm pretty sure what I described is what I've seen in the tests...

Reply:
> and it can be either side

I wonder if it is worth it to try to change this so we always get only our side. The way it currently works is confusing and can trip people up (it did me, a couple times actually).

Reply:
I put a comment on slack, it is probably worth changing it so we do not get any ElderInfo at all. our_info is already up to date we should use that.

```
            AccumulatingEvent::SectionInfo(ref info, ref key_info) => {
                let change = NeighbourChangeBuilder::new(self);
                if self.add_elders_info(info.clone(), key_info.clone(), proofs)? {
                    let change = change.build(self);
                    return Ok(Some(
                        AccumulatedEvent::new(event).with_neighbour_change(change),
                    ));
                } else {
                    return Ok(None);
                }
            }
```
We return None when caching, that event will not be processed

Reply:
Ah! "It can be either side" - so that might have been the source of the problem, I missed this at first. So yeah, we should probably change this somehow, so that we only handle our side. Then this will just work.